### PR TITLE
replace forward declarations of r2t, r3t, r4t with includes

### PR DESCRIPTION
**Pull request by hugary1995**
ref#15094

I created six headers:

RankTwoTensorForward.h
RankThreeTensorForward.h
RankFourTensorForward.h
ADRankTwoTensorForward.h
ADRankThreeTensorForward.h
ADRankFourTensorForward.h

### CODE REVIEW

**Comment by hugary1995** [comment]
I don't really understand why I need a style change. My .clang-format is the same as the one in moose root directory. Is civet using a different .clang-format? or is Atom messing up with something?

File: framework/include/utils/ADRankFourTensorForward.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ADRealForward.h"
+
+template <typename>
+class RankFourTensorTempl;
+
+typedef RankFourTensorTempl<Real> RankFourTensor;
```
Comment (lindsayad) [inline framework/include/utils/ADRankFourTensorForward.h:17]: Our goal is to reduce forward declaration duplication, so I'd suggest replacing these declarations with `#include "RankFourTensorForward.h"`. Equivalently for `ADRankThreeTensorForward.h` and `ADRankTwoTensorForward.h`

File: framework/include/utils/RankFourTensor.h [hunk 2]
```diff
@@ -442,10 +439,6 @@ struct RawType<RankFourTensorTempl<T>>
 };
 }
 
-typedef RankFourTensorTempl<Real> RankFourTensor;
-typedef RankFourTensorTempl<DualReal> DualRankFourTensor;
```
Comment (lindsayad) [inline framework/include/utils/RankFourTensor.h:446]: Go ahead and put this into the forward declaration header as well. We may still need this typedef available for backwards compatibility with applications. Equivalent comment for the other `Ranks`

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/include/utils/ADRankFourTensorForward.h (hunk 1 above)
Comment (hugary1995) [reply framework/include/utils/ADRankFourTensorForward.h]: Done.

File: framework/include/utils/RankFourTensor.h (hunk 2 above)
Comment (hugary1995) [reply framework/include/utils/RankFourTensor.h:446]: Done. The only occurrence of DualRankTwoTensor is in unit test, and I have changed that to ADRankTwoTensor. Not sure if someone should put a deprecation warning in DualRankTwoTensor.

**Review by lindsayad — APPROVED** [review]
Thanks @hugary1995 !

**Pull request merged**